sapp/ui/trace.py

```diff
@@ -403,3 +438,94 @@ def trace_kind_to_shared_text_kind(trace_kind: Optional[TraceKind]) -> SharedTex
         return SharedTextKind.SINK
 
     raise AssertionError(f"{trace_kind} is invalid")
+
+
+def to_sarif(
+    session: Session, issue: IssueQueryResult, tool: str, output_features: bool = False
+) -> SARIFCodeflowsObject:
+    postcondition_initial_frames = initial_frames(
+        session,
+        issue.issue_instance_id,
+        TraceKind.POSTCONDITION,
+    )
+    precondition_initial_frames = initial_frames(
+        session,
+        issue.issue_instance_id,
+        TraceKind.PRECONDITION,
+    )
+    postcondition_navigation = navigate_trace_frames(
+        session,
+        postcondition_initial_frames,
+        set(issue.source_kinds),
+        set(issue.sink_kinds),
+    )
+    precondition_navigation = navigate_trace_frames(
+        session,
+        precondition_initial_frames,
+        set(issue.source_kinds),
+        set(issue.sink_kinds),
+    )
+    trace_tuples = _create_trace_tuples(
+        reversed(postcondition_navigation)
+    ) + _create_trace_tuples(precondition_navigation)
+    codeflows = [{"threadFlows": [{"locations": []}]}]
+    nesting_level = 0
+    for t in trace_tuples:
+        location = _sarif_codeflow_location_from_trace_tuple(
+            t.trace_frame, tool, nesting_level, True
+        )
+        codeflows[0]["threadFlows"][0]["locations"].append(location)
+        nesting_level += 1
+
+    return codeflows
```

Comment:
Instead of creating `codeflows` to this whole structure, just use it as a list and create the structure last:

```suggestion
    codeflows = []
    nesting_level = 0
    for t in trace_tuples:
        location = _sarif_codeflow_location_from_trace_tuple(
            t.trace_frame, tool, nesting_level, True
        )
        codeflows.append(location)
        nesting_level += 1

    return [{"threadFlows": [{"locations": codeflows}]}]
```